# papers
Repository containing all scripts and documentation for published papers.

## 2018
* [Boatwright et al. 2018](http://www.genetics.org/content/early/2018/09/13/genetics.118.301564): Boatwright JL, McIntyre LM, Morse AM, Chen S, Yoo MJ, Koh J, Soltis PS, Soltis DE, Barbazuk WB. A Robust Methodology for Assessing Differential Homeolog Contributions to the Transcriptomes of Allopolyploids. GENETICS. 2018 Sept 13. 

## 2017
* [Mei et al. 2017](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5423905/): Mei W, Liu S, Schnable JC, Yeh CT, Springer NM, Schnable PS, Barbazuk WB. A Comprehensive Analysis of Alternative Splicing in Paleopolyploid Maize. Frontiers in plant science. 2017;8.

* [Feng et al. 2017](https://academic.oup.com/gbe/article/doi/10.1093/gbe/evx056/3746054/Evolution-of-the-3R-MYB-Gene-Family-in-Plants): Feng G, Burleigh JG, Braun EL, Mei W, Barbazuk WB. Evolution of the 3R-MYB Gene Family in Plants. Genome Biology and Evolution. 2017 Apr 1;9(4):1013-29.

* [Liu et al. 2017](http://onlinelibrary.wiley.com/doi/10.1111/1755-0998.12670/full): Liu X, Mei W, Soltis PS, Soltis DE, Barbazuk WB. Detecting Alternatively Spliced Transcript Isoforms from Single‐Molecule Long‐Read Sequences without a Reference Genome. Molecular Ecology Resources. 2017 Mar 1.

* [Mei et al. 2017](http://biorxiv.org/content/biorxiv/early/2017/03/24/120469.full.pdf): Mei W, Boatwright JL, Feng G, Schnable JC, Barbazuk WB. Evolutionarily Conserved Alternative Splicing Across Monocots. bioRxiv. 2017 Jan 1:120469.

